hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,346 @@
+        case_reachable(cond, case) <-- case_node(cond, i, case),
+            in_wire_value(cond, IncomingPort::from(0), v),
+            if v.supports_tag(*i);
+
+        // CFG --------------------
+        relation cfg_node(Node); // <Node> is a `CFG`
+        cfg_node(n) <-- node(n), if ctx.get_optype(*n).is_cfg();
+
+        // In `CFG` <Node>, basic block <Node> is reachable given our knowledge of predicates:
+        relation bb_reachable(Node, Node);
+        bb_reachable(cfg, entry) <-- cfg_node(cfg), if let Some(entry) = ctx.children(*cfg).next();
+        bb_reachable(cfg, bb) <-- cfg_node(cfg),
+            bb_reachable(cfg, pred),
+            output_child(pred, pred_out),
+            in_wire_value(pred_out, IncomingPort::from(0), predicate),
+            for (tag, bb) in ctx.output_neighbours(*pred).enumerate(),
+            if predicate.supports_tag(tag);
+
+        // Inputs of CFG propagate to entry block
+        out_wire_value(i_node, OutgoingPort::from(p.index()), v) <--
+            cfg_node(cfg),
+            if let Some(entry) = ctx.children(*cfg).next(),
+            input_child(entry, i_node),
+            in_wire_value(cfg, p, v);
+
+        // In `CFG` <Node>, values fed along a control-flow edge to <Node>
+        //     come out of Value outports of <Node>:
+        relation _cfg_succ_dest(Node, Node, Node);
+        _cfg_succ_dest(cfg, exit, cfg) <-- cfg_node(cfg), if let Some(exit) = ctx.children(*cfg).nth(1);
+        _cfg_succ_dest(cfg, blk, inp) <-- cfg_node(cfg),
+            for blk in ctx.children(*cfg),
+            if ctx.get_optype(blk).is_dataflow_block(),
+            input_child(blk, inp);
+
+        // Outputs of each reachable block propagated to successor block or CFG itself
+        out_wire_value(dest, OutgoingPort::from(out_p), v) <--
+            bb_reachable(cfg, pred),
+            if let Some(df_block) = ctx.get_optype(*pred).as_dataflow_block(),
+            for (succ_n, succ) in ctx.output_neighbours(*pred).enumerate(),
+            output_child(pred, out_n),
+            _cfg_succ_dest(cfg, succ, dest),
+            node_in_value_row(out_n, out_in_row),
+            if let Some(fields) = out_in_row.unpack_first(succ_n, df_block.sum_rows.get(succ_n).unwrap().len()),
+            for (out_p, v) in fields.enumerate();
+
+        // Call --------------------
+        relation func_call(Node, Node); // <Node> is a `Call` to `FuncDefn` <Node>
+        func_call(call, func_defn) <--
+            node(call),
+            if ctx.get_optype(*call).is_call(),
+            if let Some(func_defn) = ctx.static_source(*call);
+
+        out_wire_value(inp, OutgoingPort::from(p.index()), v) <--
+            func_call(call, func),
+            input_child(func, inp),
+            in_wire_value(call, p, v);
+
+        out_wire_value(call, OutgoingPort::from(p.index()), v) <--
+            func_call(call, func),
+            output_child(func, outp),
+            in_wire_value(outp, p, v);
+    };
```

Comment:
Commented that this can be done via `prepopulate_out_wire` for external functions. Whilst very basic, I think we should not do too much until we have settled how Hugr's are used as "libraries" (and perhaps also linked).

Reply:
And added `propopulate_df_inputs` as a utility. Ok to resolve?